core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregator.java

```diff
@@ -0,0 +1,271 @@
+            String fieldName, boolean filterDuplicateText,
+            Map<String, Object> metaData) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+        this.bucketCountThresholds = bucketCountThresholds;
+        this.includeExclude = includeExclude;
+        this.filterDuplicateText = filterDuplicateText;
+        this.significanceHeuristic = significanceHeuristic;
+        this.termsAggFactory = termsAggFactory;
+        this.fieldName = fieldName;
+        bucketOrds = new BytesRefHash(1, context.bigArrays());
+        dupSequenceSpotter = new DuplicateByteSequenceSpotter();        
+        lastTrieSize = dupSequenceSpotter.getEstimatedSizeInBytes();
+    }
+
+    
+    
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        final ShardGetService shardGetService = super.context.indexShard().getService();
+        final BytesRefBuilder previous = new BytesRefBuilder();
+        return new LeafBucketCollectorBase(sub, null) {
+            
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                Terms tv = ctx.reader().getTermVector(doc, fieldName);
+                if (tv!=null &&  (tv.hasPositions()  || tv.hasOffsets())){
+                    // Not necessarily faster than re-analysis in my tests!
+                    processTokenStream(doc, bucket, new TokenStreamFromTermVector(tv, -1), "");
+                    
+                } else {
+                    loadFromSource(doc, bucket, fieldName);
+                }                
+                numCollectedDocs++;
+                dupSequenceSpotter.startNewSequence();
+            }
+            
+            private void processTokenStream(int doc, long bucket, TokenStream ts, String fieldText) throws IOException{
+                if (filterDuplicateText) {
+                    ts = new DeDuplicatingTokenFilter(ts, dupSequenceSpotter);
+                }
+                CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+                ts.reset();
+                try {
+                    //Assume tokens will average 5 bytes in length to size number of tokens
+                    BytesRefHash inDocTerms = new BytesRefHash(1+(fieldText.length()/5), context.bigArrays());
+                    
+                    try{
+                        while(ts.incrementToken()){
+                            long newTrieSize = dupSequenceSpotter.getEstimatedSizeInBytes();
+                            long growth = newTrieSize - lastTrieSize;
+                            // Only update the circuitbreaker after 
+                            if(growth > MEMORY_GROWTH_REPORTING_INTERVAL_BYTES){
+                                addRequestCircuitBreakerBytes(growth);
+                                lastTrieSize = newTrieSize;
```

Comment:
It is interesting that, in true java fashion, the spotter only ever grows in size. Even when a document is done we never shrink it. This makes sense, but it is interesting.